wgpu-core/src/device/mod.rs

```diff
@@ -2355,19 +2355,40 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
                 }
             }
 
-            if let Some(start_binding) = write_map.keys().next().cloned() {
-                let descriptors = write_map
-                    .into_iter()
-                    .flat_map(|(_, list)| list)
-                    .collect::<Vec<_>>();
-                let write = hal::pso::DescriptorSetWrite {
-                    set: desc_set.raw(),
-                    binding: start_binding,
-                    array_offset: 0,
-                    descriptors,
-                };
+            if !write_map.is_empty() {
+                #[derive(PartialEq)]
+                enum DescriptorType {
+                    Buffer,
+                    Sampler,
+                    TextureView,
+                }
+                let mut writes = Vec::<hal::pso::DescriptorSetWrite<_, SmallVec<[_; 1]>>>::new();
+                let mut prev_stages = wgt::ShaderStage::empty();
+                let mut prev_ty = DescriptorType::Buffer;
+                for (binding, list) in write_map {
+                    let layout = &bind_group_layout.entries[&binding];
+                    let ty = match layout.ty {
+                        wgt::BindingType::UniformBuffer { .. } |
+                        wgt::BindingType::StorageBuffer { .. } => DescriptorType::Buffer,
+                        wgt::BindingType::Sampler { .. } => DescriptorType::Sampler,
+                        wgt::BindingType::SampledTexture { .. } |
+                        wgt::BindingType::StorageTexture { .. } => DescriptorType::TextureView,
+                    };
+                    if layout.visibility == prev_stages && ty == prev_ty {
+                        prev_stages = layout.visibility;
```

Comment:
Looks like these assignments should be under the `else` branch instead?

Reply:
Oh wow, nicely spotted! I was refactoring this branch to the opposite and forgot this piece :sweat_smile: 
Now I wonder how/why it works even